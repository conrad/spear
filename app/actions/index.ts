export * from './searchForm';
export * from './searches';
export * from "./results";
