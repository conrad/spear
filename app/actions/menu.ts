import { actionCreatorVoid } from './helpers';

export const toggleMenu = actionCreatorVoid('TOGGLE_MENU');

export function toggleMenuDropDown(): Function {
  return (dispatch: Function) => {
    dispatch(toggleMenu());
  }
}
