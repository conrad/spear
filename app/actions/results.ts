import { actionCreator, actionCreatorVoid } from './helpers';
import TextDocWriter from "../local/export/resultsExporter"; 
import * as constants from '../constants'
import IResults from '../types/IResults';

export interface IExcerptIndices {
  resultIndex: number,
  excerptIndex: number,
}

export const saveResults = actionCreator<IResults>('SAVE_RESULTS')
export const toggleShowResultsWindow = actionCreatorVoid(constants.TOGGLE_SHOW_RESULTS_WINDOW)
export const toggleShowResult = actionCreator<number>(constants.TOGGLE_SHOW_RESULT)
export const showOverlay = actionCreator<IExcerptIndices>(constants.SHOW_RESULT_OVERLAY)
export const hideOverlay = actionCreatorVoid(constants.HIDE_RESULT_OVERLAY)
export const exportResults = actionCreatorVoid(constants.EXPORT_RESULTS)

export function setResults(results: IResults) {
  return (dispatch: Function) => {
    dispatch(saveResults(results))
  }
}

export function toggleShowSearchResultRows(index: number) {
  return (dispatch: Function) => {
    dispatch(toggleShowResult(index))
  }
}

export function showResultOverlay(resultIndex: number, excerptIndex: number) {
  return (dispatch: Function) => {
    dispatch(showOverlay({ resultIndex, excerptIndex }))
  }
}

export function hideResultOverlay() {
  return (dispatch: Function) => {
    dispatch(hideOverlay())
  }
}

export function saveResultsToFile() {
  return (dispatch: Function, getState: Function) => {
    const { searches, results } = getState()
    const textDocWriter = new TextDocWriter()
    textDocWriter.saveResults("Spear_Search_Results.txt", searches.searches, results.items)
    dispatch(exportResults())
  }
}
