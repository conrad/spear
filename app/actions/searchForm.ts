import { actionCreatorVoid, actionCreator } from './helpers';

export const submitSearch = actionCreatorVoid('SUBMIT_SEARCH');
export const unsetPhrase = actionCreator<number>('DELETE_PHRASE');
export const resetFile = actionCreatorVoid('RESET_FILE');
export const setFile = actionCreator<string>('SET_FILE');
export const appendPhrase = actionCreator<string>('APPEND_PHRASE');
export const setAddingPhrase = actionCreator<boolean>('SET_ADDING_PHRASE');

export function addFile(filename: string) {
  return (dispatch: Function, getState: Function) => {
    // const { searchForm } = getState();
    dispatch(setFile(filename));
  }
}

export function addPhrase(phrase: string) {
  return (dispatch: Function, getState: Function) => {
    // const { searchForm } = getState();
    console.log('addphrase in actions');
    dispatch(appendPhrase(phrase));
    dispatch(setAddingPhrase(false));
  }
}

export function deletePhrase(index: number) {
  return (dispatch: Function, getState: Function) => {
    // const { searchForm } = getState();
    dispatch(unsetPhrase(index));
  }
}

export function startAddingPhrase() {
  return (dispatch: Function) => {
    dispatch(setAddingPhrase(true));
  };
}

export function submitValidSearch() {
  return (dispatch: Function, getState: Function) => {
    const { searchForm } = getState();

    if (!searchForm.filename || searchForm.phrases.length < 1) {
      return;
    }

    dispatch(submitSearch());
  };
}
