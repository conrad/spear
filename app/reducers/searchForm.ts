import { IAction } from '../actions/helpers';
import { submitSearch, appendPhrase, unsetPhrase, setFile, resetFile } from '../actions/searchForm';
import { clone } from './helpers';


const initialState = { 
  filename: "",
  phrases: ['hey ma', 'whats up']   
};

export interface IFormState {
  filename: string,
  phrases: Array<string>
};

export type TState = IFormState;

export default function searchForm(state: IFormState = initialState, action: IAction) {
  let newState: IFormState = clone(state);

  if (appendPhrase.test(action)) {
    newState.phrases.push(action.payload);
    console.log('appendPhrase in reducer. payload:', action.payload, 'state:', state);
    return newState;
  } else if (unsetPhrase.test(action)) {
    newState.phrases.splice(action.payload, 1);
    return newState;
  } else if (setFile.test(action)) {
    newState.filename = action.payload;
    return newState;
  } else if (resetFile.test(action)) {
    newState.filename = '';
    return newState;
  } else if (submitSearch.test(action)) {
    console.log("filename:", newState.filename);
    console.log("first phrase:", newState.phrases[0]);
    return newState;
  }

  return newState;
}
