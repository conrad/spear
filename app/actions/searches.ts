import { actionCreator, actionCreatorVoid } from './helpers';
import { IPhrase, ISearch, IMove } from '../reducers/searches';
import JsonReader from '../local/import/searchLoader';
import JsonWriter from '../local/export/searchExporter';

export const storeSearch = actionCreator<ISearch>('STORE_SEARCH');
export const deleteSearch = actionCreator<ISearch>('DELETE_SEARCH');
export const moveSearch = actionCreator<IMove>('MOVE_SEARCH');
export const unsetPhrase = actionCreator<IPhrase>('UNSET_PHRASE');
export const toggleSearchAsUsed = actionCreator<number>('TOGGLE_SEARCH_AS_USED');
export const setNewSearchName = actionCreator<string>('SET_NEW_SEARCH_NAME');
export const storeNewSearch = actionCreatorVoid('STORE_NEW_SEARCH');
export const setActiveSearch = actionCreator<number>('SET_ACTIVE_SEARCH');
export const storeSearchesFromProfile = actionCreator<Array<ISearch>>('STORE_SEARCHES_FROM_PROFILE');
export const exportSearchesToFile = actionCreatorVoid('EXPORT_SEARCHES_TO_FILE') 

export function addSearch(search: ISearch) {
  return (dispatch: Function) => {
    dispatch(storeSearch(search));
  }
}

export function updateSearch(search: ISearch) {
  return (dispatch: Function) => {
    dispatch(storeSearch(search));
  }
}

export function reorderSearch(move: IMove) {
  return (dispatch: Function) => {
    dispatch(moveSearch(move));
  };
}

export function deletePhrase(phraseIndex: number, searchIndex: number) {
  return (dispatch: Function) => {
    const phrase: IPhrase = {
      index: phraseIndex,
      text: '',
      searchIndex,
    };
    
    dispatch(unsetPhrase(phrase));
  }    
}

export function setSearchAsUsed(index: number, isUsed: boolean|null) {
  return (dispatch: Function) => {
    dispatch(toggleSearchAsUsed(index));
  };
}

export function updateNewSearchName(name: string) {
  return (dispatch: Function) => {
    dispatch(setNewSearchName(name));
  };    
}

export function exportSearches() {
  return (dispatch: Function, getState: Function) => {
    const { searches } = getState();
    const jsonWriter = new JsonWriter();
    try {
      jsonWriter.saveProfile('searches_profile.json', searches);
    } catch (e) {
      console.log('Error saving searches:', e);
    }
    // dispatch(exportSearchesToFile());
  }
}

export function addSearchesFromProfile(file: File) {
  const searches: Array<ISearch> = JsonReader.retrieveSearchesFromFile(file.path);

  return (dispatch: Function) => {
    dispatch(storeSearchesFromProfile(searches));
  };    
}

export function addNewSearch() {
  return (dispatch: Function) => {
    dispatch(storeNewSearch());
  };     
}

export function selectSearch(index: number) {
  return (dispatch: Function) => {
    dispatch(setActiveSearch(index));
  };     
}

export function removeSearch(index: number, name: string) {
  return (dispatch: Function) => {
    const search = {
      index,
      name,
      phrases: [],
      isIncluded: false,
      isEditing: false
    };
    dispatch(deleteSearch(search));
  };
}
