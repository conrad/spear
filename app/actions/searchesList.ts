import { actionCreator } from './helpers';
import { ISearch, IMove } from '../reducers/searchesList';

export const storeSearch = actionCreator<ISearch>('STORE_SEARCH');
export const deleteSearch = actionCreator<ISearch>('DELETE_SEARCH');
export const moveSearch = actionCreator<IMove>('MOVE_SEARCH');

export function addSearch(index: number, name: string, description: string|null, phrases: Array<string>) {
  return (dispatch: Function) => {
    const search = {
      index,
      name,
      description,
      phrases,
      isIncluded: false,
      isEditing: false,
    };

    dispatch(storeSearch(search));
  }
}

export function reorderSearch(move: IMove) {
  return (dispatch: Function) => {
    dispatch(moveSearch(move));
  };
}

export function removeSearch(index: number, name: string) {
  return (dispatch: Function) => {
    const search = {
      index,
      name,
      description: '',
      phrases: [],
      isIncluded: false,
      isEditing: false
    };
    dispatch(deleteSearch(search));
  };
}
