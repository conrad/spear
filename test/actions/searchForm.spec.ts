import { spy } from 'sinon';
import * as actions from '../../app/actions/searchForm';

describe('actions', () => {
  it('should submitSearch should create submitSearch action', () => {
    expect(actions.submitSearch()).toMatchSnapshot();
  });

  it('should resetFile should create resetFile action', () => {
    expect(actions.resetFile()).toMatchSnapshot();
  });

  it('should unsetPhrase should create unsetPhrase action', () => {
    expect(actions.unsetPhrase(1)).toMatchSnapshot();
  });
  
  it('should setFile should create setFile action', () => {
    expect(actions.setFile('name-for-a-file')).toMatchSnapshot();
  });
  
  it('should appendPhrase should create appendPhrase action', () => {
    expect(actions.appendPhrase('something to look for')).toMatchSnapshot();
  });
  
  it('should setNewPhrase should create setNewPhrase action', () => {
    expect(actions.setNewPhrase('something to look for')).toMatchSnapshot();
  });

  it('should setIsPhraseUsed should create setIsPhraseUsed action', () => {
    expect(actions.setIsPhraseUsed(false)).toMatchSnapshot();
  });

  it('should updateNewPhrase should create setNewPhrase action', () => {
    let newPhrase: string = 'new phrase text';
    const fn = actions.updateNewPhrase(newPhrase);
    expect(fn).toBeInstanceOf(Function);
    const dispatch = spy();
    fn(dispatch);
    expect(dispatch.calledWith({ 
      type: actions.setNewPhrase.type, 
      payload: newPhrase
    })).toBe(true);
  });

  it('should updateIsNewPhraseUsed should create setIsPhraseUsed action to true', () => {
    const fn = actions.updateIsNewPhraseUsed(true);
    expect(fn).toBeInstanceOf(Function);
    const dispatch = spy();
    fn(dispatch);
    expect(dispatch.calledWith({ 
      type: actions.setIsPhraseUsed.type, 
      payload: true
    })).toBe(true);
  });

  it('should updateIsNewPhraseUsed should create setIsPhraseUsed action to false', () => {
    const fn = actions.updateIsNewPhraseUsed(false);
    expect(fn).toBeInstanceOf(Function);
    const dispatch = spy();
    fn(dispatch);
    expect(dispatch.calledWith({ 
      type: actions.setIsPhraseUsed.type, 
      payload: false
    })).toBe(true);
  });
  
  it('should addFile should create setFile action', () => {
    const filename: string = 'thenameofthefile.txt';
    const fn = actions.addFile(filename);
    expect(fn).toBeInstanceOf(Function);
    const dispatch = spy();
    fn(dispatch);
    expect(dispatch.calledWith({ 
      type: actions.setFile.type, 
      payload: filename
    })).toBe(true);
  });

  it('should addPhrase should create appendPhrase action', () => {
    const phrase: string = 'some particular phrase';
    const fn = actions.addPhrase(phrase);
    expect(fn).toBeInstanceOf(Function);
    const dispatch = spy();
    fn(dispatch);
    expect(dispatch.calledWith({ 
      type: actions.appendPhrase.type, 
      payload: phrase
    })).toBe(true);
  });

  it('should deletePhrase should create unsetPhrase action', () => {
    const index: number = 3;
    const fn = actions.deletePhrase(index);
    expect(fn).toBeInstanceOf(Function);
    const dispatch = spy();
    fn(dispatch);
    expect(dispatch.calledWith({ 
      type: actions.unsetPhrase.type, 
      payload: index
    })).toBe(true);
  });

  it('should submitValidSearch should create submitSearch action', () => {
    const fn = actions.submitSearchIfValid();
    expect(fn).toBeInstanceOf(Function);
    const dispatch = spy();
    const getState = () => ({ 
      searchForm: {
        filename: 'hello.txt',
        phrases: ['one', 'two']
      }
    });
    fn(dispatch, getState);
    expect(dispatch.calledWith({ type: actions.submitSearch.type })).toBe(true);
  });

  it('should submitValidSearch should not create submitSearch action when filename is missing', () => {
    const fn = actions.submitSearchIfValid();
    expect(fn).toBeInstanceOf(Function);
    const dispatch = spy();
    const getState = () => ({ 
      searchForm: {
        filename: '',
        phrases: ['one', 'two']
      }
    });
    fn(dispatch, getState);
    expect(dispatch.notCalled).toBe(true);
  });

  it('should submitValidSearch should not create submitSearch action when no phrases are included in search', () => {
    const fn = actions.submitSearchIfValid();
    expect(fn).toBeInstanceOf(Function);
    const dispatch = spy();
    const getState = () => ({ 
      searchForm: {
        filename: 'hello.txt',
        phrases: []
      }
    });
    fn(dispatch, getState);
    expect(dispatch.notCalled).toBe(true);
  });
});