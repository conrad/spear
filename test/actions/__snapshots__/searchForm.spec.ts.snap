// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`actions should resetFile should create resetFile action 1`] = `
Object {
  "type": "RESET_FILE",
}
`;

exports[`actions should setFile should create setFile action 1`] = `
Object {
  "payload": Object {
    "lastModifiedDate": 0,
    "msClose": [Function],
    "msDetachStream": [Function],
    "name": "thenameofthefile.txt",
    "path": "/all/the/way.txt",
    "size": 10,
    "slice": [Function],
    "type": "gorillaz/clintEastwood",
    "webkitRelativePath": "",
  },
  "type": "SET_FILE",
}
`;

exports[`actions should setIsPhraseUsed should create setIsPhraseUsed action 1`] = `
Object {
  "payload": false,
  "type": "SET_IS_PHRASE_USED",
}
`;

exports[`actions should setNewPhrase should create setNewPhrase action 1`] = `
Object {
  "payload": "something to look for",
  "type": "SET_NEW_PHRASE",
}
`;

exports[`actions should submitSearch should create submitSearch action 1`] = `
Object {
  "type": "SUBMIT_SEARCH",
}
`;

exports[`searchForm actions should resetFile should create resetFile action 1`] = `
Object {
  "type": "RESET_FILE",
}
`;

exports[`searchForm actions should setFile should create setFile action 1`] = `
Object {
  "payload": Object {
    "lastModified": 1,
    "lastModifiedDate": 2018-08-04T23:44:12.251Z,
    "msClose": [Function],
    "msDetachStream": [Function],
    "name": "thenameofthefile.txt",
    "path": "/all/the/way.txt",
    "size": 10,
    "slice": [Function],
    "type": "gorillaz/clintEastwood",
    "webkitRelativePath": "",
  },
  "type": "SET_FILE",
}
`;

exports[`searchForm actions should setIsPhraseUsed should create setIsPhraseUsed action 1`] = `
Object {
  "payload": false,
  "type": "SET_IS_PHRASE_USED",
}
`;

exports[`searchForm actions should setNewPhrase should create setNewPhrase action 1`] = `
Object {
  "payload": "something to look for",
  "type": "SET_NEW_PHRASE",
}
`;

exports[`searchForm actions should submitSearch should create submitSearch action 1`] = `
Object {
  "type": "SUBMIT_SEARCH",
}
`;
